refs #11737 - connect to localhost for qpid

stbenjam opened the pull request:
Required PR's
- [ ] https://github.com/Katello/katello-installer/pull/260 (installer migration)
- [ ] puppet-katello (this one)
- [ ] https://github.com/Katello/puppet-katello_devel/pull/67
- [ ] https://github.com/Katello/puppet-capsule/pull/96
- [ ] https://github.com/Katello/puppet-certs/pull/65
- [ ] https://github.com/Katello/puppet-candlepin/pull/44

Test with:

  `./setup.rb --module-prs katello/79,capsule/55,gutterball/17,certs/65`

Existing katello needs upgrade and certs update:

```
  foreman-installer --scenario katello --certs-update-all --upgrade
```

stbenjam commented:
I need to test the installer PR still, is there any effect from scenarios on this? I know Kafo now has some kind of migration support, no?  I've been out of the loop for a bit.

stbenjam commented:
Can hold off on testing this, there's some issue where the cert isn't getting updated on qpid itself. :-(

```
Feb 22 12:28:30 katello-centos7.example.com pulp[32178]: pulp.server.async.scheduler:ERROR: ("Connection hostname 'localhost' does not match names from peer certificate: ['katello-centos7.example.com', u'katello-centos7.example.com']",)
```

It's correct in the ssl-build directory, but the nssdb thingy doesn't seem to get the update.

ehelms commented:
There is now migration supported directly by kafo, take a look at the following for a full example:

https://github.com/Katello/katello-installer/pull/290/files#diff-adbb84ad728dfced62b0e8f2d90b4503

stbenjam commented:
Great, thanks, updated the installer PR - and tested both a new install and an existing.  This should be good to review/test, thanks!

ehelms commented:
Will Capsule's need an upgrade/update with this change?

stbenjam commented:
They do, their qpid will also only listen on localhost, and the pulp workers connect to it.  That should be covered here, but I guess they'll need a new certificate bundle generated with the `localhost` cname.

ekohl commented on manifests/init.pp:
```diff
@@ -102,18 +102,19 @@
     ssl_cert_db            => $::certs::nss_db_dir,
     ssl_cert_password_file => $::certs::qpid::nss_db_password_file,
     ssl_cert_name          => 'broker',
+    interface              => 'lo',
```
Am I missing something or is this unused?

stbenjam replied:
It's used. 

https://github.com/Katello/puppet-qpid/blob/master/manifests/init.pp#L18
https://github.com/Katello/puppet-qpid/blob/master/templates/qpidd.conf.erb#L42

ekohl replied:
That's puppet-qpid, this is puppet-katello. I don't see it being passed on to that module.

stbenjam replied:
You're commenting on a qpid class resource.

ekohl replied:
Right, so I was missing something. Not sure why I thought this was a class parameter ```katello``` instead of ```qpid```. My bad.

ehelms commented:
This may need a rebase? I tried this out and got the following, but I was also using an updated way of doing module-pr testing with kdeploy (https://github.com/Katello/katello-deploy/pull/213):

```
==> qpid: Debug: hiera(): Looking up apache::mod::passenger:: 0 0.0/s 00:00:05
==> qpid: Debug: hiera(): Looking up candlepin::db_name in Y: 0 0.0/s 00:00:06
==> qpid:  Puppet::Parser::AST::Resource failed with error ArgumentError: Could not find declared class ::certs::qpid_client at /usr/share/katello-installer-base/modules/katello/manifests/init.pp:107 on node qpid.example.com
==> qpid: 
==> qpid:  Puppet::Parser::AST::Resource failed with error ArgumentError: Could not find declared class ::certs::qpid_client at /usr/share/katello-installer-base/modules/katello/manifests/init.pp:107 on node qpid.example.com
```

stbenjam commented:
That error doesn't look like any code I've changed.  I just rebased all my PR's but they would've rebased cleanly if you wanted to see if that was the problem. I don't see what changes the katello-deploy would've caused this either.

This need better coordination.  I tested this a few weeks ago and it was fine, but who knows what happened since then. I'm reluctant to go through the whole process of testing it myself and figuring outs what's wrong if it'll just sit for more weeks once I come back with an answer.

stbenjam commented:
Ok, tested with new katello-deploy PR after rebasing and it worked. :100: 

I ran through a test after rebasing the modules and that seems to have been the problem even though there were no conflicts.  

If you're too busy to come back to this in the next day or two, ping me a little bit beforehand so I can make sure it's all still current so we avoid playing more back and forth.


```
Launching installer with command: foreman-installer --scenario katello 

Running shell command: foreman-installer --scenario katello 
Installing             Done                                               [100%] [...............................................................................................................................]
  Success!
  * Katello is running at https://centos7-bats.example.com
      Initial credentials are admin / u8j98mUCkdWdKDPL
  * To install additional capsule on separate machine continue by running:

      capsule-certs-generate --capsule-fqdn "$CAPSULE" --certs-tar "~/$CAPSULE-certs.tar"

  The full log is at /var/log/foreman-installer/katello.log
[root@centos7-bats vagrant]# netstat -anp | grep qpid
tcp        0      0 127.0.0.1:5671          0.0.0.0:*               LISTEN      14552/qpidd         
```

ehelms commented:
I am keeping my setup in tact, and ran through the test once you rebased. I hit an error with candlepin, however it looks like the typical timeout that sometimes occurs so I will re-run and ping back.

ehelms commented:
Hrmm failed again with the below, I will dig into the logs, but figured I'd post in case you thought of something.  My box config is:

```
qpid:
  box: centos7-katello-nightly
  options: --module-prs katello/79,capsule/55,gutterball/17,certs/65
```

```
==> qpid:  /usr/bin/wget --no-proxy --timeout=30 --tries=40 --wait=20 --retry-connrefused -qO- http://localhost:8080/candlepin/admin/init > /var/log/candlepin/cpinit.log 2>&1 && touch /var/lib/candlepin/cpinit_done returned 8 instead of one of [0]
==> qpid: Error: /usr/bin/wget --no-proxy --timeout=30 --tri: 787/788, 99%, 0.0/s, elapsed: 00:06:33
==> qpid:  /Stage[main]/Candlepin::Service/Exec[cpinit]/returns: change from notrun to 0 failed: /usr/bin/wget --no-proxy --timeout=30 --tries=40 --wait=20 --retry-connrefused -qO- http://localhost:8080/candlepin/admin/init > /var/log/candlepin/cpinit.log 2>&1 && touch /var/lib/candlepin/cpinit_done returned 8 instead of one of [0]
==> qpid: 
==> qpid: Debug: Executing '/bin/systemctl is-enabled httpd': 787/788, 99%, 0.0/s, elapsed: 00:06:34
==> qpid: Notice: /Stage[main]/Apache::Service/Service[httpd: 787/788, 99%, 0.0/s, elapsed: 00:06:35
==> qpid: Notice: /Stage[main]/Foreman::Database/Foreman::Ra: 787/788, 99%, 0.0/s, elapsed: 00:06:48
==> qpid:  /Stage[main]/Foreman::Database/Foreman::Rake[db:seed]/Exec[foreman-rake-db:seed]: Failed to call refresh: /usr/sbin/foreman-rake db:seed returned 1 instead of one of [0]
==> qpid:  /Stage[main]/Foreman::Database/Foreman::Rake[db:seed]/Exec[foreman-rake-db:seed]: /usr/sbin/foreman-rake db:seed returned 1 instead of one of [0]
```

stbenjam commented:
Hmm, I used a clean centos7-bats box.  Was this box already installed before, and maybe some of the certs didn't get the localhost CN?


You might want to try this and see if it fixes it:

    foreman-installer --scenario katello --certs-update-all --upgrade

stbenjam commented:
Came up fine for me again, how are you passing install options to centos7-nightly? Doesn't it immediately jump into an install as part of "vagrant up"?

stbenjam commented:
Tried the method you were using and it came up fine, when you try again and it fails can I get a foreman-debug?

stbenjam commented:
PR's went out of sync, rebased again.

Still getting a successful install:

```
==> qpid: Done                                              : 793/793, 100%, 0.0/s, elapsed: 00:11:19
==> qpid:   Success!
==> qpid:   * Katello is running at https://qpid.example.com
==> qpid:       Initial credentials are admin / VXWwxPdUjDQMHW27
==> qpid:   * To install additional capsule on separate machine continue by running:
==> qpid: 
==> qpid:       capsule-certs-generate --capsule-fqdn "$CAPSULE" --certs-tar "~/$CAPSULE-certs.tar"
==> qpid: 
==> qpid:   The full log is at /var/log/foreman-installer/katello.log
```

Can I get a foreman-debug next time you try?

stbenjam commented:
Turns out my vagrant setup is somehow creating this /etc/hosts record: `127.0.0.1   qpid.example.com`

I missed an AMQP reference in candlepin: https://github.com/Katello/puppet-candlepin/pull/44

But I was getting successful installs since my FQDN resolved to 127.0.0.1, where qpid was listening.

This should do it now:

`./setup.rb --module-prs katello/79,capsule/55,gutterball/17,certs/65,candlepin/44`

stbenjam commented:
Candlepin (well, the qpid library) only verifies the main CN - it can't verify alternate DNS hostnames. Filed https://bugzilla.redhat.com/show_bug.cgi?id=1329327

stbenjam commented:
Got around to testing this again, but the fix to candlepin doesn't seem to work: https://bugzilla.redhat.com/show_bug.cgi?id=1329327#c8

stbenjam commented:
Candlepin is fixed, and in koji (candlepin-0.9.54.10-1). 

Everything seems to be working:

```
tcp        0      0 127.0.0.1:5671          0.0.0.0:*               LISTEN      27339/qpidd         
```

```
[root@qpid-localhost ~]# hammer ping
candlepin:      
    Status:          ok
    Server Response: Duration: 26ms
candlepin_auth: 
    Status:          ok
    Server Response: Duration: 24ms
pulp:           
    Status:          ok
    Server Response: Duration: 55ms
foreman_tasks:  
    Status:          ok
    Server Response: Duration: 698ms
```
